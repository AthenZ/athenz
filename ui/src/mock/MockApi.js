const origApi = require('../api').default;

export default class MockApi {
    static setMockApi(mockApi) {
        require('../api').default = () => {
            return mockApi;
        };
    }
    static cleanMockApi() {
        require('../api').default = () => {
            return origApi();
        };
    }
}
