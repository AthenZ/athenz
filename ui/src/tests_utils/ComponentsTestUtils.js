/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { initStore } from '../redux/store';
import { getExpiryTime } from '../redux/utils';

export const renderWithRedux = (component, initialState = {}) => {
    return render(
        <Provider store={initStore(initialState)}>{component}</Provider>
    );
};

export const buildUserForState = (user) => {
    return {
        expiry: getExpiryTime(),
        user,
    };
};

export const getStateWithUser = (user, initialState = {}) => {
    return { ...initialState, user };
};

export const buildDomainDataForState = (domainData, domainName = 'dom') => {
    return {
        domainName,
        expiry: getExpiryTime(),
        domainData,
    };
};

export const getStateWithDomainData = (domainData, initialState = {}) => {
    return { ...initialState, domainData: domainData };
};

export const getStateWithRoles = (roles, initialState = {}) => {
    return { ...initialState, roles };
};

export const buildRolesForState = (roles, domainName = 'dom') => {
    return {
        domainName,
        expiry: getExpiryTime(),
        roles,
    };
};

export const buildPoliciesForState = (policies, domainName = 'dom') => {
    return {
        domainName,
        expiry: getExpiryTime(),
        policies,
    };
};

export const getStateWithPolicies = (policies, initialState = {}) => {
    return { ...initialState, policies };
};

export const buildServicesForState = (
    services,
    domainName = 'dom',
    allProviders = []
) => {
    return {
        domainName,
        expiry: getExpiryTime(),
        services,
        allProviders,
    };
};

export const getStateWithServices = (services, initialState = {}) => {
    return { ...initialState, services };
};

export const buildMicrosegmentationForState = (
    inboundOutboundList,
    domainName = 'dom'
) => {
    return {
        domainName,
        expiry: getExpiryTime(),
        inboundOutboundList,
    };
};

export const getStateWithMicrosegmentation = (
    microsegmentation,
    initialState = {}
) => {
    return { ...initialState, microsegmentation };
};

export const buildGroupsForState = (groups, domainName = 'dom') => {
    return {
        domainName,
        expiry: getExpiryTime(),
        groups,
    };
};

export const getStateWithGroups = (groups, initialState = {}) => {
    return { ...initialState, groups };
};

export const buildServiceDependencies = (
    serviceDependencies,
    domainName = 'dom'
) => {
    return {
        domainName,
        expiry: getExpiryTime(),
        serviceDependencies,
    };
};

export const getStateWithServiceDependencies = (
    serviceDependencies,
    initialState = {}
) => {
    return { ...initialState, serviceDependencies: serviceDependencies };
};

export const mockRolesApiCalls = () => {
    return {
        getRoles: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve([]);
            })
        ),
        getRoleMembers: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve({});
            })
        ),
    };
};

export const mockPoliciesApiCalls = () => {
    return {
        getPolicies: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve([]);
            })
        ),
        getPolicy: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve({});
            })
        ),
        // getRoleMembers: jest.fn().mockReturnValue(
        //     new Promise((resolve, reject) => {
        //         resolve({});
        //     })
        // ),
    };
};

export const mockAllDomainDataApiCalls = (domainDetails, headerDetails) => {
    return {
        getDomain: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve(domainDetails || {});
            })
        ),
        isAWSTemplateApplied: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve('');
            })
        ),
        getHeaderDetails: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve(headerDetails || {});
            })
        ),
        getPendingDomainMembersListByDomain: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve([]);
            })
        ),
        getFeatureFlag: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve('');
            })
        ),
        getMeta: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve([]);
            })
        ),
        getAuthorityAttributes: jest.fn().mockReturnValue(
            new Promise((resolve, reject) => {
                resolve({});
            })
        ),
        getReviewGroups: jest.fn().mockReturnValue([]),
        getReviewRoles: jest.fn().mockReturnValue([]),
        getPageFeatureFlag: jest.fn().mockResolvedValue({}),
    };
};

export const getStateWithUserList = (userList, initialState = {}) => {
    return { ...initialState, user: userList };
};

export const getStateWithDomainDataAndUserList = (
    domainData,
    userList,
    initialState = {}
) => {
    return { ...initialState, domainData: domainData, user: userList };
};
