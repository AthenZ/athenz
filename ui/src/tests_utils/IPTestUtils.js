/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export default class IPTestUtils {
    static getRandomInt(min, max) {
        return Math.floor(Math.random() * (max - min + 1)) + min;
    }

    static generateRandomIP() {
        return `${IPTestUtils.getRandomInt(0, 255)}.${IPTestUtils.getRandomInt(
            0,
            255
        )}.${IPTestUtils.getRandomInt(0, 255)}.${IPTestUtils.getRandomInt(
            0,
            255
        )}`;
    }

    static generateRandomCIDR() {
        const ip = IPTestUtils.generateRandomIP();
        const prefix = IPTestUtils.getRandomInt(0, 32);
        return `${ip}/${prefix}`;
    }

    static generateRandomString(length) {
        const characters =
            'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
        let result = '';
        for (let i = 0; i < length; i++) {
            result += characters.charAt(
                Math.floor(Math.random() * characters.length)
            );
        }
        return result;
    }

    static generateRandomFQDN() {
        const subdomain = IPTestUtils.generateRandomString(
            IPTestUtils.getRandomInt(3, 10)
        );
        const domain = IPTestUtils.generateRandomString(
            IPTestUtils.getRandomInt(3, 10)
        );
        const tld = IPTestUtils.generateRandomString(
            IPTestUtils.getRandomInt(2, 6)
        ); // Top Level Domain, like com, net, org, etc.
        return `${subdomain}.${domain}.${tld}`;
    }
}
