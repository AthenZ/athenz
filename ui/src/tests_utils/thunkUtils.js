/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import {
    addAssertionPolicyVersionToStore,
    deleteAssertionPolicyVersionFromStore,
    loadPolicies,
} from '../redux/actions/policies';
import { getExpiryTime } from '../redux/utils';
import { storeDomainData, storePolicies } from '../redux/actions/domains';
import { loadDomainData } from '../redux/actions/domain-data';

export const getLoadDomainDataAction = (domainName, domainData) => {
    return loadDomainData(domainData, domainName, getExpiryTime());
};

export const getStoreDomainDataAction = (domainData) => {
    return storeDomainData(domainData);
};

export const getLoadPoliciesAction = (domainName, policies = {}) => {
    return loadPolicies(policies, domainName, getExpiryTime());
};

export const getStorePoliciesAction = (policiesData) => {
    return storePolicies(policiesData);
};

export const getDeleteAssertionPolicyVersionAction = (
    policyName,
    version,
    assertionId
) => {
    return deleteAssertionPolicyVersionFromStore(
        policyName,
        version,
        assertionId
    );
};

export const getAddAssertionPolicyVersionAction = (
    policyName,
    version,
    newAssertion
) => {
    return addAssertionPolicyVersionToStore(policyName, version, newAssertion);
};
