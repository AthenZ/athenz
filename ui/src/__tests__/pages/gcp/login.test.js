/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { waitFor } from '@testing-library/react';
import { renderWithRedux } from '../../../tests_utils/ComponentsTestUtils';
import GCPLoginPage from '../../../pages/gcp/login';
import MockApi from '../../../mock/MockApi';

afterEach(() => {
    MockApi.cleanMockApi();
});

describe('GCP Login Page', () => {
    it('should render error no gcp project when api returns no results', async () => {
        const mockApi = {
            getResourceAccessList: jest.fn().mockReturnValue([]),
        };
        MockApi.setMockApi(mockApi);
        const { getByTestId } = renderWithRedux(<GCPLoginPage />, {
            isFetching: false,
        });
        await waitFor(() =>
            expect(getByTestId('gcp-login-error')).toMatchSnapshot()
        );
    });

    it('should render projects api returns results', async () => {
        let testData = {
            resources: [
                {
                    assertions: [
                        {
                            role: 'dummy.project:role.user.dev',
                        },
                    ],
                },
            ],
        };
        const mockApi = {
            getResourceAccessList: jest
                .fn()
                .mockReturnValue(Promise.resolve(testData)),
        };
        MockApi.setMockApi(mockApi);
        const { getByTestId } = renderWithRedux(<GCPLoginPage />);
        await waitFor(() => expect(getByTestId('gcp-login')).toMatchSnapshot());
    });
});
