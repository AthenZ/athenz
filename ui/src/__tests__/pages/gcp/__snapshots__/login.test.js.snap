// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GCP Login Page should render error no gcp project when api returns no results 1`] = `
.emotion-0 {
  border-bottom: 1px solid lightgray;
  margin: 10px;
  margin-bottom: 20px;
}

.emotion-2 {
  max-width: 800px;
  margin: 0 auto;
}

<div
  data-testid="gcp-login-error"
>
  <header
    class="emotion-0 emotion-1"
  >
    <img
      src="/static/google-cloud.svg"
    />
  </header>
  <div
    class="emotion-2 emotion-3"
  >
    <h3>
      Error: There are no GCP project roles associated with your account.
    </h3>
    <p>
      Check to make sure that your 
      <tt>
        gcp.*
      </tt>
       
      roles contain your user.
      <br />
      E.g., if your account is 'jdoe', then
       
      <tt>
        gcp.fed.admin.user
      </tt>
       should have
       
      <tt>
        user
        .jdoe
      </tt>
       as a member.
    </p>
    <p>
      If that does not work, please ask your Athenz domain admin for assistance.
    </p>
  </div>
</div>
`;

exports[`GCP Login Page should render projects api returns results 1`] = `
.emotion-0 {
  border-bottom: 1px solid lightgray;
  margin: 10px;
  margin-bottom: 20px;
}

.emotion-0 {
  border-bottom: 1px solid lightgray;
  margin: 10px;
  margin-bottom: 20px;
}

.emotion-2 {
  max-width: 800px;
  margin: 0 auto;
}

.emotion-2 {
  max-width: 800px;
  margin: 0 auto;
}

.emotion-4 {
  font-size: 16px;
}

.emotion-6 {
  font-size: 18px;
  display: inline;
  border-bottom: 1px solid lightgray;
  margin-bottom: 10px;
  padding: 10px;
  width: 100%;
  display: block;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-top: 10px;
  margin-bottom: 10px;
  margin-left: 20px;
  font-weight: bold;
}

.emotion-10 {
  vertical-align: middle;
  width: 18px;
  height: 18px;
  margin: 10px;
}

.emotion-12 {
  margin: 0;
  font-weight: bold;
  font-size: 16px;
  color: #000;
  vertical-align: middle;
}

.emotion-14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-16 {
  border: none;
  border-radius: 2px;
  box-shadow: 0 0 0 1px transparent inset,0 0 0 0 #d5d5d5 inset;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  font-family: Helvetica,Arial,sans-serif;
  font-weight: 300;
  font-size: 14px;
  line-height: 1;
  margin: 5px;
  outline: 0;
  padding: 10px 24px;
  text-align: center;
  text-shadow: none;
  text-transform: none;
  -webkit-transition: all 0.2s ease-in;
  transition: all 0.2s ease-in;
  vertical-align: baseline;
  white-space: nowrap;
  background: linear-gradient(
              to right,
              #3697f2,
              #3570f4
          );
  color: #fff;
}

.emotion-16:first-of-type {
  margin-left: 0;
}

.emotion-16:disabled {
  background: rgba(48,48,48,0.1);
  color: rgba(48,48,48,0.2);
  cursor: not-allowed;
}

.emotion-16:hover:not(:disabled) {
  background: linear-gradient(
                  to right,
                  #3570f4,
                  #3448f7
              );
}

<div
  data-testid="gcp-login"
>
  <header
    class="emotion-0 emotion-1"
  >
    <img
      src="/static/google-cloud.svg"
    />
  </header>
  <div
    class="emotion-2 emotion-3"
  >
    <form
      action="/gcp/login/post"
      method="post"
    >
      <input
        name="_csrf"
        type="hidden"
        value=""
      />
      <input
        name="isAdmin"
        type="hidden"
        value=""
      />
      <p
        class="emotion-4 emotion-5"
      >
        Select a role:
      </p>
      <div>
        <div
          class="emotion-6 emotion-7"
        >
          Project: 
          dummy-project-id
           (
          dummy.project
          )
        </div>
        <div
          class="emotion-8 emotion-9"
        >
          <div>
            <input
              class="emotion-10 emotion-11"
              name="project"
              required=""
              type="radio"
              value="dummy.project:role.user.dev"
            />
            <label
              class="emotion-12 emotion-13"
            >
              user.dev
            </label>
          </div>
        </div>
      </div>
      <div
        class="emotion-14 emotion-15"
      >
        <button
          class="emotion-16 denali-button"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;
